clientapi/auth/login_token.go

```diff
@@ -0,0 +1,209 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/clientapi/userutil"
+	"github.com/matrix-org/dendrite/internal/config"
+	"github.com/matrix-org/dendrite/userapi/api"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/util"
+)
+
+// This file handles all the m.login.token logic
+
+// GetAccountByLocalpart function implemented by the appropriate database type
+type GetAccountByLocalpart func(ctx context.Context, localpart string) (*api.Account, error)
+
+// LoginTokenRequest struct to hold the possible parameters from an m.login.token http request
+type LoginTokenRequest struct {
+	Login
+	Token string `json:"token"`
+	TxnID string `json:"txn_id"`
+}
+
+// LoginTypeToken holds the configs and the appropriate GetAccountByLocalpart function for the database
+type LoginTypeToken struct {
+	GetAccountByLocalpart GetAccountByLocalpart
+	Config                *config.ClientAPI
+}
+
+// Name returns the expected type of "m.login.token"
+func (t *LoginTypeToken) Name() string {
+	return "m.login.token"
+}
+
+// Request returns a struct of type LoginTokenRequest
+func (t *LoginTypeToken) Request() interface{} {
+	return &LoginTokenRequest{}
+}
+
+// Type of the LoginToken
+type loginToken struct {
+	UserID       string
+	CreationTime int64
+	RandomPart   string
+}
+
+// Login completes the whole token validation, user verification for m.login.token
+// returns a struct of type *auth.Login which has the users details
+func (t *LoginTypeToken) Login(ctx context.Context, req interface{}) (*Login, *util.JSONResponse) {
+	r := req.(*LoginTokenRequest)
+	userID, err := validateLoginToken(r.Token, r.TxnID, &t.Config.Matrix.ServerName)
+	if err != nil {
+		return nil, &util.JSONResponse{
+			Code: http.StatusUnauthorized,
+			JSON: jsonerror.InvalidArgumentValue(err.Error()),
+		}
+	}
+	r.Login.Identifier.User = userID
+	r.Login.Identifier.Type = "m.id.user"
+
+	return &r.Login, nil
+}
+
+// Decodes and validates a LoginToken
+// Accepts the base64 encoded token string as param
+// Checks the time expiry, userID (only the format, doesn't check to see if the user exists)
+// Also checks the DB to see if the token exists
+// Returns the localpart if successful
+func validateLoginToken(tokenStr string, txnID string, serverName *gomatrixserverlib.ServerName) (string, error) {
+	token, err := decodeLoginToken(tokenStr)
+	if err != nil {
+		return "", err
+	}
+
+	// check whether the token has a valid time.
+	// TODO: should this 5 second window be configurable?
+	if time.Now().Unix()-token.CreationTime > 5 {
```

Comment:
5 seconds seems very short, perhaps 1 minute?

Reply:
I only choose 5 seconds because the spec suggested it: https://matrix.org/docs/spec/client_server/r0.6.1#handling-the-authentication-endpoint

Reply:
Oh wow I see. In which case please just link to this on this line to justify why 5s.

Reply:
5 seconds sounds pretty short to me as well, if you have a connection or DNS failure that could easily time out. It seems that Synapse uses 2 minutes, although I was unable to find any reference to why that was chosen: https://github.com/matrix-org/synapse/blob/78d5f91de1a9baf4dbb0a794cb49a799f29f7a38/synapse/handlers/auth.py#L1323-L1325

Reply:
In that case, should I make this time configurable by the admin (with a default time of 5 seconds)?